### DISCUSSION TITLE
Adding Amiberry support to AmigaVision

### DISCUSSION THREAD

**Discussion by limi** [BlitterStudio/amiberry#1350] [General]
Hello! I am one of the maintainers of the [AmigaVision](https://amiga.vision) project, and we aim to take out the complexity and frustration out of Amiga emulation. The project started on MiSTer, also works on original hardware (A1200/A4000/CD32), as well as Analogue Pocket FPGA and other FPGA boards like MiST and SiDi.

We recently started to look into seriously supporting emulators, and it’s been… a journey. UAE configs are extremely confusing, and when you layer things like WinUAE/FS-UAE/Amiberry/PUAE on top, it gets messy, and I still don’t have an emulator setup that lives up to the standard we are trying to provide.

High level, here’s what we are trying to accomplish (from the end-user side) — I don’t know if Amiberry can do this, but here’s what we’re aiming for:

- Configurations that we want to provide should support:
  - **TV setup**: 1920×1080px full screen, 16:9 (will by extension also work on 4K TVs)
  - **Handheld support**: 640×480 full screen, 4:3 (this is the most common resolution on emulation handhelds)
- Integer scaling.
- Crop-to-active-pixels.
- Scanlines and/or filters/shaders (but if the implementation is to just make every other line black, we’ll stay with nearest-neighbor integer scaling)
- The ability to double click a settings file on the desktop, and (as long as you have the emulator installed), it opens up with our configuration settings, including HDF setups, and all the other settings already mentioned. No UI should be shown.
  - As an example, this already works great on FS-UAE — you double click `AmigaVision.fs-uae`, and the emulator loads straight into the setup.

Is there anything in this list that seems hard to implement with an Amiberry setup? I have hit some walls/bugs around trying to request 1080p via either fullscreen or “fake fullscreen” via a window, but I don’t know enough to file useful bug reports yet. No matter what I specify of the combinations, it either ends up at the native resolution of the display, or falls back to 800×600px in the settings.

Here’s what I have tried so far in `amiberry.conf`:
```
default_correct_aspect_ratio=yes
default_auto_crop=yes
default_width=1920
default_height=1080
default_fullscreen_mode=1
Quickstart=0
```

**Comment by giantclambake** [comment]
I actually have a MegaAGS setup here ~ typically I go for fullscreen, but in this [YT video](https://www.youtube.com/watch?v=ts3pYJcP_Vk) I've opted to show windowed mode. I could've double-clicked on the icon, but chose right-click, execute.... for the longhand effect ;) I have tested on teevee, but alas I really dislike Amiga titles + widescreen aspect ratios...usually speaking, that's part of one of my x86-64 thin-client setups, connected to 19" 4:3 display (with builtin CRT modes), and it works best there (I consider 16:9 aspect ratio somewhat 'blasphemous' wrt properly rendering Amiga titles to scene ... workbench being the exception of course =)

I honestly have no idea what the config.uae looks like ~ I created it with the amiberry GUI ....it's not very often I find the need to futz with amiberry.conf directly. I'll include it for reference's sake. Probably I just wanted to more or less demonstrate your last point above, as being entirely possible.

I do know this however...fs-uae 3.x is a very different thing to amiberry, in many different ways =)

Obviously, midwan will have to weigh in on the technical questions.

[MegaAGS.uae.gz](https://github.com/user-attachments/files/15766171/MegaAGS.uae.gz)

**Reply by limi** [reply]
> I actually have a MegaAGS setup here

Great! Happy to see you are enjoying our work 😄 

Can I recommend updating from the 2021 version? So, so many improvements since then!

>  ~ typically I go for fullscreen, but in this [YT video](https://www.youtube.com/watch?v=ts3pYJcP_Vk) I've opted to show windowed mode.

Will check it out! Windowed mode works fine, but we are looking for a good setup that works on modern TVs, which means fullscreen 1080p.

> I could've double-clicked on the icon, but chose right-click, execute.... for the longhand effect ;) 

I certainly can’t double click any config file in the Mac version of Amiberry to launch it with that configuration — what are you launching in this case?

> I have tested on teevee, but alas I really dislike Amiga titles + widescreen aspect ratios […] (I consider 16:9 aspect ratio somewhat 'blasphemous' wrt properly rendering Amiga titles to scene ... workbench being the exception of course =)

You just need to do it while respecting the Pixel Aspect Ratios and have horizontal/vertical offset and scaling settings for each game 😄 

This is a partial write up of how we do per-game settings for MiSTer, 3000+ hand-tuned settings: https://amiga.vision/5x

Will take a look at your video + settings file, this definitely helps, thanks!

**Comment by midwan** [comment]
Jumping straight to the bullet points:
> * Configurations that we want to provide should support:
>   
>   * **TV setup**: 1920×1080px full screen, 16:9 (will by extension also work on 4K TVs)
>   * **Handheld support**: 640×480 full screen, 4:3 (this is the most common resolution on emulation handhelds)

I've found the best results to be either using Full-window or Windowed mode. Fullscreen mode under SDL2 sometimes gave... unpredictable results. However, Full-window mode does mean that we'll be using the current resolution the host has, and the GPU will do the scaling to the full size of it. I haven't seen a problem with this approach, except with scenarios where the refresh rate is less than the emulated one (e.g. on 4K TVs which might be doing 25Hz instead of 50Hz, as is the case with the RPI4 by default). Obviously this would work fine with 640x480 as well, and the output will be scaled accordingly automatically.

> * Integer scaling.

I don't have Integer Scaling as an option implemented yet. Only Aspect Ratio correction, which is enabled by default. Maybe in a future update :)

> * Crop-to-active-pixels.

This is available as the "Auto-Crop" function. It can be enabled by default (from `amiberry.conf`) or manually, or even mapped to a hotkey and triggered on-demand (in which case it will apply once, when triggered, not always). The alternative is to use Manual Crop, with H/V offset values, to fine-tune your output. But honestly, after the updates I made in Auto-Crop lately, it seems like the most reasonable choice now. :)

> * Scanlines and/or filters/shaders (but if the implementation is to just make every other line black, we’ll stay with nearest-neighbor integer scaling)

There is no shader support in Amiberry yet. This was planned with an OpenGL back-end, which has not really materialized yet (there is some experimental code, but it's still buggy). There is Scaline support built-in, but it's not as good as a proper shader. It might be "good enough" for your needs. :)

> * The ability to double click a settings file on the desktop, and (as long as you have the emulator installed), it opens up with our configuration settings, including HDF setups, and all the other settings already mentioned. No UI should be shown.
>   

This is do-able as well, since you can associate Amiberry as the default handler of such files. The alternative would be to have a shortcut that runs a script, which executes the emulator along with any optional parameters you may want. The later approach was usually done in older versions of Amiberry, but with the recent updates to handle filetypes directly as a parameter, it should not be necessary.

>   * As an example, this already works great on FS-UAE — you double click `AmigaVision.fs-uae`, and the emulator loads straight into the setup.
> 
> Is there anything in this list that seems hard to implement with an Amiberry setup? I have hit some walls/bugs around trying to request 1080p via either fullscreen or “fake fullscreen” via a window, but I don’t know enough to file useful bug reports yet. No matter what I specify of the combinations, it either ends up at the native resolution of the display, or falls back to 800×600px in the settings.
> 
> Here’s what I have tried so far in `amiberry.conf`:
> 
> ```
> default_correct_aspect_ratio=yes
> default_auto_crop=yes
> default_width=1920
> default_height=1080
> default_fullscreen_mode=1
> Quickstart=0
> ```

See my comments about using Full-Window above, instead of Fullscreen. Keep in mind that if you ask for Auto Crop to be enabled, the actual image output will no longer be the Amiga native one, but smaller (depending on the active pixels). It will still be scaled keeping the aspect ratio, of course.

Also, the `default_width` and `default_height` values here should indicate the desired Amiga video mode output, not native display. The default values are the full width and height (e.g. 720x576), so you cannot go above those values - it will be ignored. You can set it to smaller values, if you wanted to crop some pixels by default - usually in this case you would also want to have H/V Centering enabled also, otherwise the image might appear shifted.

**Reply by midwan** [reply]
@limi 
You might want to hop in the Discord server also, if you want: https://discord.gg/KuwvkmqU
